shared/subspace-data-retrieval/src/piece_getter.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (C) 2024 Subspace Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Getting object pieces from the Subspace Distributed Storage Network, or various caches.
+
+use async_trait::async_trait;
+use std::fmt;
+use std::sync::Arc;
+use subspace_archiving::archiver::NewArchivedSegment;
+use subspace_core_primitives::{Piece, PieceIndex};
+
+/// A type-erased error
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+/// Object piece getter errors.
+#[derive(Debug, thiserror::Error)]
+pub enum PieceGetterError {
+    /// Getting piece failed, a retry won't get the piece from this provider, try another provider
```

Comment:
There is no provider notion in this API. Why not returning `None` if piece can't be found just like API already allows to express?

Reply:
```suggestion
    /// Getting piece failed, a retry won't get the piece from this piece getter, try a different piece getter
```

Because `Ok(None)` is temporary and can retry, but `Err(_)` is fatal?
>    /// Returns `Ok(None)` for temporary errors: the piece is not found, but immediately retrying
>    /// this provider might return it.
>    /// Returns `Err(_)` for permanent errors: this provider can't provide the piece at this time,
>    /// and another provider should be attempted.

Happy to change that to something else, is there another error handing scheme you’d prefer?

Reply:
My preference would be for piece getter to do retries internally if it needs to. If it returns an error, it is an error that called has to deal with (like reconstructing it from segment pieces).

I also don't expect developers to have multiple piece getters to choose from and if they they'd compose them, so one way or another I'm not sure this is a very helpful comment, but I don't have a strong preference here.

Reply:
Sure, happy to update the docs and implementation to do this instead